src/zenml/orchestrators/utils.py

```diff
@@ -305,3 +308,95 @@ def _get_model_versions_from_artifacts(
+    will be restored.
+    """
+
+    def __init__(self, target_artifact_store_id: Optional[UUID]) -> None:
+        """Initialization of the context manager.
+
+        Args:
+            target_artifact_store_id: the ID of the artifact store to load.
+        """
+        self.active_artifact_store_id = Client().active_stack.artifact_store.id
+        self.target_artifact_store_id = target_artifact_store_id
+
+    def __enter__(self) -> "BaseArtifactStore":
+        """Instantiate the target artifact store and return it.
+
+        Returns:
+            the python object
+        """
+        try:
+            if self.target_artifact_store_id is not None:
+                if (
+                    self.active_artifact_store_id
+                    != self.target_artifact_store_id
+                ):
+                    get_logger(__name__).warning(
+                        f"Trying to use the artifact store with ID:"
+                        f"'{self.target_artifact_store_id}'"
+                        f"which is currently not the active artifact store."
+                    )
+
+                artifact_store_model_response = Client().get_stack_component(
+                    component_type=StackComponentType.ARTIFACT_STORE,
+                    name_id_or_prefix=self.target_artifact_store_id,
+                )
+                return cast(
+                    "BaseArtifactStore",
+                    StackComponent.from_model(artifact_store_model_response),
+                )
+            else:
+                return Client().active_stack.artifact_store
+
+        except KeyError:
+            raise RuntimeError(
+                "Unable to fetch the artifact store with id: "
+                f"'{self.target_artifact_store_id}'. Check whether the artifact"
+                "store still exists and you have to right permissions to "
+                "access it."
+            )
+        except ImportError:
+            raise RuntimeError(
+                "Unable to load the implementation of the artifact store with"
+                f"id: '{self.target_artifact_store_id}'. Please make sure that "
+                "the environment that you are loading this artifact from "
+                "has the right dependencies."
+            )
+
+    def __exit__(
+        self,
+        exc_type: Optional[Any],
+        exc_value: Optional[Any],
+        traceback: Optional[Any],
+    ) -> None:
+        """Set it back to the original state.
+
+        Args:
+            exc_type: The class of the exception
+            exc_value: The instance of the exception
+            traceback: The traceback of the exception
+        """
+        active_artifact_store_response = Client().get_stack_component(
+            component_type=StackComponentType.ARTIFACT_STORE,
+            name_id_or_prefix=self.active_artifact_store_id,
+        )
+        _ = StackComponent.from_model(active_artifact_store_response)
```

Comment:
Why do we need this? Is this a hack to overwrite the creds in fileio? If so, maybe we can have something cleaner, since this is just calls with some unexpected byproduct, IMO.

Reply:
The goal here is to recreate the active artifact store, which in turn registers the correct filesystem (with the proper set of credentials) in the filesystem registry. I agree that the behavior might not be very obvious at first glance but I would not categorize it as a hack though.

Reply:
I think a comment to explain would be nice at least. Or maybe instead of storing the artifact store ID, just store the object and call `self.artifact_store._register()` instead?

Reply:
I added a comment on why we do this. As for the `__exit__`, I switched to `Client().active_stack..artifact_store._register()` instead.

Reply:
Is it at all possible for the `Client().active_stack.artifact_store` to change between the `__enter__` and `__exit__` @schustmi ?

Reply:
Very unlikely but sure it's possible. But actually you probably want to register the active artifact store in that case right, not the old one?

Reply:
Makes sense. It's a good thing that's exactly what the code does right now then 😄.